### ISSUE TITLE
Summary / Check your answers component is missing

### ISSUE THREAD

**Issue by opensourcegeek** [UKHomeOffice/govuk-react#466]
It would be good to do a Summary page / Check your answers sort of component within govuk-react. It could be simple styled components using dl/dt/dd etc like shown [here](https://govuk-prototype-kit.herokuapp.com/docs/templates/check-your-answers). I've a very basic component like that working, and we're using it in few places now. I do see benefit in building it once in this library. What are your thoughts on that?

**Comment by penx** [comment]
We’re not intending to track the prototype kit directly at the moment but could make some exceptions if it looks production ready. 

Does this component exist in the govuk design system, design system backlog or govuk-frontend?

If you can link to the sass that you propose we should track then even better

**Comment by opensourcegeek** [comment]
@penx I don't think it appears as a separate component itself, however it exists as a [pattern](https://design-system.service.gov.uk/patterns/check-answers/) in govuk design system. If this pattern can be achieved with existing components exposed then an example would be really helpful as it wasn't clear how to go about it. 

So we've hand rolled this pattern currently with our own set of styled components that wrap `dl`, `dt` and `dd` as in the example. I'd really like to stay close to this project so that either we could help you guys where possible and also make use of all good work from this project instead of reinventing it all ourselves.

**Comment by penx** [comment]
Backlog url 
https://github.com/alphagov/govuk-design-system-backlog/issues/36

**Comment by penx** [comment]
Prototype template
https://github.com/alphagov/govuk-prototype-kit/blob/master/docs/views/templates/check-your-answers.html

**Comment by penx** [comment]
Prototype sass
https://github.com/alphagov/govuk-prototype-kit/blob/master/app/assets/sass/patterns/_check-your-answers.scss

**Comment by penx** [comment]
I’d be keen to have a component in govuk-react that:
- is called CheckYourAnswers
- uses the Sass linked above and documents reasons for any deviation
- outputs HTML that matches the above nunjucks template

I would recommend the proposed JSX usage is defined here first.

**Comment by penx** [comment]
WIP proposal for JSX usage

```jsx
<CheckYourAnswers>
  <CheckYourAnswers.Item
    question={<CheckYourAnswers.Question>{dom content}</CheckYourAnswers.Question>}
    answer={<CheckYourAnswers.Answer>{dom content}</CheckYourAnswers.Answer>}
    onChange={function}
    changeUrl={string}
  />
</CheckYourAnswers>

**Comment by opensourcegeek** [comment]
@penx Makes sense - the way I'd done this has a separate element for each level as shown below

```jsx
      <SummaryTable>
        <SummaryRow>
          <SummaryCellKey>Some Key</SummaryCellKey>
          <SummaryCellValue>{someValue}</SummaryCellValue>
          <SummaryCellLink>
             <ChangeLinkTag to={'/add-some-url'}>Change</ChangeLinkTag>
          </SummaryCellLink>
        </SummaryRow>
      </SummaryTable>  
```
I'm happy to refactor it to align with your proposed API. The key above is question and value is answer and there's another link element which explicitly maps to a react-router `Link` for now but that could be changed to take either `Link` or `a`. One thing I'd come across is, when the data structure you're exposing is flat then this (key, value) pair kind of idea works really well. However when you have nested data structure like mapping one to many relationships it's not clear how to do the layout for clearly separating this parent child relationship. 

We tried to group each nested element together without any underlining, this looks visually readable and understandable that nested item with it's own key value pairs is an attribute of parent. If you can think of better ways to group them I'd be happy to get that implemented as well at least within our project with your new proposed API. I did wonder about using table for showing the nested data structure but it doesn't scale very well for smaller screens and mixing them with normal summary page components it just looks awkward.

**Comment by penx** [comment]
I’d want to exactly follow the markup and css used by govuk prototype kit unless there was good reason not to. 

We don’t want to introduce react router as a dependency, but instead allow an ‘as’ prop for a react router Link to be passed in. Maybe here it would be called changeAs

**Comment by opensourcegeek** [comment]
@penx I'll look into any differences tonight. Just skimming through I can see one difference where we have wrapped this "change" link with `span` prototype kit uses `dd`. But CSS for both are very similar. If they're producing exactly same markup and CSS would you like a PR with your proposed API on top?

**Comment by nickcolley** [comment]
Just FYI we're planning on doing this by building a new component, the work in progress is here: https://github.com/alphagov/govuk-frontend/pull/1065

**Comment by penx** [comment]
Great, lets wait til this is in govuk-frontend master before replicating it here

**Comment by opensourcegeek** [comment]
@nickcolley @penx Thanks

**Comment by jsrobertson** [comment]
Has anyone started looking at this yet?  If not, I can take a look at some point over the next couple of weeks as it's something that would benefit us a lot - https://design-system.service.gov.uk/components/summary-list/

**Comment by penx** [comment]
@jsrobertson sorry for the late reply 😄

in case you hadn't guessed the answer was no.

Feel free to assign this to yourself if you still think this is something you could pick up.

**Comment by Vaent** [comment]
I will likely be creating a [summary list](https://design-system.service.gov.uk/components/summary-list/) component in the near future for our govuk-react-based frontend.

If I can do it in a generic enough way which still works for our use case, I'll look to raise a pull request here as well.

If there's any existing work on this component which I could pick up and build on, that would be great!

**Comment by Vaent** [comment]
The [alphagov summary list Nunjucks template](https://github.com/alphagov/govuk-frontend/pull/1065/files#diff-264540801816da205b27070f314283adf1f6971befb20c94921add08997f6802) supports rows structured as `| key | value | ( [...actions] | )?`

Item 1: mandatory key as `<dt>`
Item 2: mandatory value as `<dd>`
Item 3: optional list of actions as `<ul>` wrapped in a single `<dd>`; excluded from HTML if no actions present

All items may be passed as HTML or text. For actions, the HTML/text content is wrapped in an `<a>` tag and may optionally specify visually hidden text to follow the content.

Would we expect a govuk-react component to fully match this? My use case only requires a single optional action after the key & value - supporting a list would be an extension piece. I guess I'll start with what I need and from there assess how useful it's likely to be.

Edit: in contrast to the above, the HTML example for the [GDS check your answers pattern](https://design-system.service.gov.uk/patterns/check-answers/) doesn't bother with a list, the action is a simple singular link in the final `<dd>` tag of each row.
